api/app/geospatial/wps_dataset.py

```diff
@@ -0,0 +1,202 @@
+from typing import Optional
+from osgeo import gdal, osr
+import numpy as np
+
+from app.utils.geospatial import GDALResamplingMethod
+
+
+class WPSDataset:
+    """
+    A wrapper around gdal datasets for common operations
+    """
+
+    def __init__(self, ds_path: Optional[str], ds=None, band=1, chunk_size=256, access=gdal.GA_ReadOnly, datatype=gdal.GDT_Byte):
+        self.ds = ds
+        self.ds_path = ds_path
+        self.band = band
+        self.chunk_size = chunk_size
+        self.access = access
+        self.datatype = datatype
+
+    def __enter__(self):
+        if self.ds is None:
+            self.ds: gdal.Dataset = gdal.Open(self.ds_path, self.access)
```

Comment:
Since ds_path is optional, does anything bad happen if gdal.Open is called with None?

Reply:
We'll just have `None` for `self.ds` at that point, any further operations will fail.